src/backend/src/plugins/telemetry/telemetry_service_impl.h

```diff
@@ -1,669 +1,1774 @@
+
+        rpc_obj->set_allocated_acceleration_frd(
+            translateToRpcAccelerationFrd(imu.acceleration_frd).release());
+
+        rpc_obj->set_allocated_angular_velocity_frd(
+            translateToRpcAngularVelocityFrd(imu.angular_velocity_frd).release());
+
+        rpc_obj->set_allocated_magnetic_field_frd(
+            translateToRpcMagneticFieldFrd(imu.magnetic_field_frd).release());
+
+        rpc_obj->set_temperature_degc(imu.temperature_degc);
+
+        return rpc_obj;
+    }
+
+    static rpc::telemetry::TelemetryResult::Result
+    translateToRpcResult(const mavsdk::Telemetry::Result& result)
+    {
+        switch (result) {
+            default:
+                LogErr() << "Unknown result enum value: " << static_cast<int>(result);
+            // FALLTHROUGH
+            case mavsdk::Telemetry::Result::Unknown:
+                return rpc::telemetry::TelemetryResult_Result_RESULT_UNKNOWN;
+            case mavsdk::Telemetry::Result::Success:
+                return rpc::telemetry::TelemetryResult_Result_RESULT_SUCCESS;
+            case mavsdk::Telemetry::Result::NoSystem:
+                return rpc::telemetry::TelemetryResult_Result_RESULT_NO_SYSTEM;
+            case mavsdk::Telemetry::Result::ConnectionError:
+                return rpc::telemetry::TelemetryResult_Result_RESULT_CONNECTION_ERROR;
+            case mavsdk::Telemetry::Result::Busy:
+                return rpc::telemetry::TelemetryResult_Result_RESULT_BUSY;
+            case mavsdk::Telemetry::Result::CommandDenied:
+                return rpc::telemetry::TelemetryResult_Result_RESULT_COMMAND_DENIED;
+            case mavsdk::Telemetry::Result::Timeout:
+                return rpc::telemetry::TelemetryResult_Result_RESULT_TIMEOUT;
+        }
+    }
+
+    grpc::Status SubscribePosition(
+        grpc::ServerContext* /* context */,
+        const mavsdk::rpc::telemetry::SubscribePositionRequest* /* request */,
+        grpc::ServerWriter<rpc::telemetry::PositionResponse>* writer) override
+    {
+        auto stream_closed_promise = std::make_shared<std::promise<void>>();
+        auto stream_closed_future = stream_closed_promise->get_future();
+        register_stream_stop_promise(stream_closed_promise);
+
+        auto is_finished = std::make_shared<bool>(false);
+
+        std::mutex subscribe_mutex{};
+
+        _telemetry.position_async(
+            [this, &writer, &stream_closed_promise, is_finished, &subscribe_mutex](
+                const mavsdk::Telemetry::Position& position) {
+                rpc::telemetry::PositionResponse rpc_response;
+
+                rpc_response.set_allocated_position(translateToRpcPosition(position).release());
+
+                std::lock_guard<std::mutex> lock(subscribe_mutex);
+                if (!*is_finished && !writer->Write(rpc_response)) {
+                    _telemetry.position_async(nullptr);
+                    *is_finished = true;
+                    stream_closed_promise->set_value();
```

Comment:
I think if you `stop()` here, the promise goes out of scope and the `weak_ptr` to stop it is not valid anymore. However, it is a leak in the `vector` with invalid `weak_ptr`s.

Reply:
I correct that. It would be a race during the time where the promise is not yet out of scope but already set. So it needs an `unregister` as you suggest.

Reply:
Great!